### PR TITLE
Modifief kexec_load syscall to support running EFI application.

### PR DESCRIPTION
**Pull request by oweisse**
Currently only supports extremely simple apps like HelloWorld.efi

* kexec.h - Added KEXEC_RUN_PE flag to be passed to kexec_load. This is the
* kexec.c
        - Implemented kimage_load_pe, kimage_run_pe
        - Implemented EFI hooks to generate fake EFI system table and
        fake EFI boot services.
        - Implemented naive EFI AllocatePool, FreePool
        - Implemented EFI conout to enable OutputString

### CODE REVIEW

File: kernel/kexec.c [hunk 1]
```diff
@@ -103,6 +104,782 @@ static int kimage_alloc_init(struct kimage **rimage, unsigned long entry,
...
+                }
+        }
+
+        if( i % 8 != 0 )
+                printk( KERN_ERR  "%s\n", output);
+
+        printk( KERN_ERR  "\n");
+}
+
+/* This implementationis based on kimage_load_normal_segment */
```
Comment (hugelgupf) [inline kernel/kexec.c:160]: how is this different from kimage_load_normal_segment?

File: kernel/kexec.c [hunk 2]
```diff
@@ -103,6 +104,782 @@ static int kimage_alloc_init(struct kimage **rimage, unsigned long entry,
...
+        EfiConventionalMemory,
+        EfiUnusableMemory,
+        EfiACPIReclaimMemory,
+        EfiACPIMemoryNVS,
+        EfiMemoryMappedIO,
+        EfiMemoryMappedIOPortSpace,
+        EfiPalCode,
+        EfiPersistentMemory,
+        EfiMaxMemoryType
+} EFI_MEMORY_TYPE;
```
Comment (hugelgupf) [inline kernel/kexec.c]: Can you dump these in a separate file?

File: kernel/kexec.c [hunk 3]
```diff
@@ -103,6 +104,782 @@ static int kimage_alloc_init(struct kimage **rimage, unsigned long entry,
...
+		maddr  += mchunk;
+		buf    += mchunk;
+		mbytes -= mchunk;
+	}
+
+	return result;
+}
+
+
+void kimage_load_pe(struct kimage *image, unsigned long nr_segments)
```
Comment (hugelgupf) [inline kernel/kexec.c]: how is this different from the normal segment loading? one contiguous chunk instead of each segment ending up somewhere random?

File: kernel/kexec.c (hunk 1 above)
Comment (oweisse) [reply kernel/kexec.c:160]: We took inspiration from kimage_load_normal_segment, but our implementation is quite different. 

For example, kimage_load_normal_segment allocates memory on the fly in potentially non-continuous memory. We assume a continuous piece of memory was already allocated. 
Additionally, kimage_load_normal_segment has different operation depending on image->file_mode whereas this flag is not relevant for us. 

Long story short, if we would re-use kimage_load_normal_segment we would complicate it substaintially.

File: kernel/kexec.c (hunk 2 above)
Comment (oweisse) [reply kernel/kexec.c]: Moved to linux/efi.h

File: kernel/kexec.c (hunk 3 above)
Comment (oweisse) [reply kernel/kexec.c]: Yes,
1. One contiguous chunk
2. Patching image according to relocations table (the last segment)
3. Preparing EFI_LOADED_IMAGE_PROTOCOL struct with the entry point
4. Allocating memory in a precise location which is ImageBase. This is theoretically un necessary but it made our lives easier. If nothing else, debugging with gdb is now easier becuase we have a constant loading address.
